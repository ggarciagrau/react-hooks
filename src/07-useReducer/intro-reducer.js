const initialState = [
    {
        id: 1,
        todo: "Ir al gym",
        done: false
    },
    {
        id: 2,
        todo: "Compra mercadona",
        done: false
    }
];

const todoReducer = (state = initialState, action = {}) => {

    if (action.type == 'ADD_TODO') {
        return [...state, action.payload];
    }

    return state;
}

let todos = todoReducer();

console.log(todos)

const newTodo = {
    id: 3,
    todo: "Ser un pro de React",
    done: false
}

todos = todoReducer(todos, {
    type: 'ADD_TODO',
    payload: newTodo
});

console.log(todos)