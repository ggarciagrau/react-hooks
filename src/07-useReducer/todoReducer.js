export const todoReducer = (state, { type, payload }) => {

    switch (type) {
        case "ADD_TODO":
            return [
                ...state, {
                    id: new Date().getTime(),
                    description: payload,
                    done: false
                }
            ];
        default:
            return state;
    }
}