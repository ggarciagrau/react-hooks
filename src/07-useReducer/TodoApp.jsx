import { useTodo } from "../hooks/useTodo";
import { TodoList } from "./components/TodoList";
import { TodoItem } from "./components/TodoItem";
import { TodoForm } from "./components/TodoForm";

export const TodoApp = () => {

    const {
        todos,
        handleNewTodo,
        handleRemoveTodo,
        toggleTodo
    } = useTodo();

    return (
        <>
            <h1>Todo App x, <small>pendientes: x</small></h1>
            <hr />

            <div className="row">
                <div className="col-7">
                    <TodoList>
                        {
                            todos.map(({ id, description, done }) =>
                                <TodoItem
                                    key={id}
                                    {...{
                                        id,
                                        description,
                                        checked: done,
                                        handleRemoveTodo,
                                        toggleTodo
                                    }}
                                />
                            )
                        }
                    </TodoList>
                </div>

                <div className="col-5">
                    <h4>Agregar TODO</h4>
                    <hr />

                    <TodoForm handleNewTodo={handleNewTodo} />
                </div>
            </div>
        </>
    );
}