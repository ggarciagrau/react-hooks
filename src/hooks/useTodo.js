import { todoReducer } from "../07-useReducer/todoReducer";
import { useReducer, useEffect } from "react";

export const useTodo = (initialState) => {

    const init = () => JSON.parse(localStorage.getItem('todos')) || [];

    const [state, dispatch] = useReducer(todoReducer, initialState, init);

    const nPending = state.filter(todo => todo.done != true).length;

    useEffect(() => {

        localStorage.setItem('todos', JSON.stringify(state || []));

    }, [state]);

    const handleNewTodo = description => {

        dispatch({ type: "ADD_TODO", payload: description });
    }

    const handleRemoveTodo = id => {

        dispatch({ type: "REMOVE_TODO", payload: id });
    }

    const toggleTodo = id => {

        dispatch({ type: "TOGGLE_TODO", payload: id });
    }

    return {
        todos: state,
        nPending,
        handleNewTodo,
        handleRemoveTodo,
        toggleTodo
    }

}