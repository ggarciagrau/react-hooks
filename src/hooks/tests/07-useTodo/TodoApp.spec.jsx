import { render } from "@testing-library/react";
import { TodoApp } from "../../../07-useReducer/TodoApp";
import { useTodo } from "../../useTodo";

const todosList = [
  {
    id: 1,
    description: "Canary",
    done: false,
  },
  {
    id: 2,
    description: "Budgie",
    done: false,
  },
];

const mockHandleNewTodo = jest.fn();
const mockRemoveTodo = jest.fn();
const mockToggleTodo = jest.fn();

jest.mock("../../useTodo");

useTodo.mockReturnValue({
  todos: todosList,
  nPending: 2,
  handleNewTodo: mockHandleNewTodo,
  handleRemoveTodo: mockRemoveTodo,
  toggleTodo: mockToggleTodo,
});

describe("Tests for TodoApp", () => {
  it("Should display the component by default", () => {
    const rendered = render(<TodoApp />);

    expect(rendered.getByText("Todo App 2,"))
    expect(rendered.getByText("pendientes: 2"))
    expect(rendered.container).toMatchSnapshot();
  });
});
