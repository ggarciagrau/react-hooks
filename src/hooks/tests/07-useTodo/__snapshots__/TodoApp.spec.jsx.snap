// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for TodoApp Should display the component by default 1`] = `
<div>
  <h1>
    Todo App 
    2
    , 
    <small>
      pendientes: 
      2
    </small>
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-7"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex justify-content-between"
        >
          <span
            class="align-self-center"
          >
            Canary
          </span>
          <input
            type="checkbox"
          />
          <button
            class="btn btn-danger"
          >
            Delete
          </button>
        </li>
        <li
          class="list-group-item d-flex justify-content-between"
        >
          <span
            class="align-self-center"
          >
            Budgie
          </span>
          <input
            type="checkbox"
          />
          <button
            class="btn btn-danger"
          >
            Delete
          </button>
        </li>
      </ul>
    </div>
    <div
      class="col-5"
    >
      <h4>
        Agregar TODO
      </h4>
      <hr />
      <form>
        <input
          class="form-control"
          name="description"
          placeholder="New todo"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-2"
          type="submit"
        >
          Add
        </button>
      </form>
    </div>
  </div>
</div>
`;
